packer/grapl-service/image.pkr.hcl

```diff
@@ -0,0 +1,163 @@
+  type    = string
+  default = "v0.9.1"
+}
+
+variable "git_sha" {
+  description = "The git SHA of the commit the AMI is being generated from. If present, will be used to tag the AMI."
+  type        = string
+  default     = env("GIT_SHA")
+}
+
+variable "git_branch" {
+  description = "The git branch the AMI is being generated from. If present, will be used to tag the AMI."
+  type        = string
+  default     = env("GIT_BRANCH")
+}
+
+variable "buildkite_build_number" {
+  description = "The build number of the Buildkite pipeline this AMI was built in. If present, will be used to tag the AMI."
+  type        = string
+  default     = env("BUILDKITE_BUILD_NUMBER")
+}
+
+locals {
+  # These are various metadata tags we can add to the resulting
+  # AMI. If any are unset (like the Buildkite build number, if
+  # building outside of Buildkite), those tags will be filtered out
+  # and not added to the AMI.
+  tags_from_vars = {
+    "git:sha"                = "${var.git_sha}"
+    "git:branch"             = "${var.git_branch}"
+    "buildkite:build_number" = "${var.buildkite_build_number}"
+  }
+}
+
+locals {
+  formatted_timestamp = formatdate("YYYYMMDDhhmmss", timestamp())
+  copy_ami_to_regions = [
+    "us-east-2",
+    "us-west-1",
+    "us-west-2",
+  ]
+}
+
+data "amazon-ami" "amazon-linux-2-x86_64" {
+  filters = {
+    architecture                       = "x86_64"
+    "block-device-mapping.volume-type" = "gp2"
+    name                               = "*amzn2-ami-hvm-*"
+    root-device-type                   = "ebs"
+    virtualization-type                = "hvm"
+  }
+  most_recent = true
+  owners      = ["amazon"]
+  region      = "${var.aws_region}"
+}
+
+source "amazon-ebs" "amazon-linux-2-amd64-ami" {
+  ami_description = "An Amazon Linux 2 x86_64 AMI that has Nomad and Consul installed."
+  ami_name        = "${var.ami_name_prefix}-amazon-linux-2-amd64-${local.formatted_timestamp}"
+  instance_type   = "t2.micro"
+  region          = "${var.aws_region}"
+  # Where we copy it after it's built
+  ami_regions  = local.copy_ami_to_regions
+  source_ami   = data.amazon-ami.amazon-linux-2-x86_64.id
+  ssh_username = "ec2-user"
+
+  tag {
+    key   = "BuiltBy"
+    value = "Packer ${packer.version}"
+  }
+
+  # Filter out any variable-based tags that are unset.
+  dynamic "tag" {
+    for_each = { for key, value in local.tags_from_vars : key => value if value != "" }
+    content {
+      key   = tag.key
+      value = tag.value
+    }
+  }
+
+  skip_create_ami = "${var.build_ami == true ? false : true}"
+}
+
+build {
+  sources = ["source.amazon-ebs.amazon-linux-2-amd64-ami"]
+
+  provisioner "shell" {
+    inline = [
+      "sudo yum install -y git",
+      # As recommended in https://github.com/hashicorp/terraform-aws-nomad/tree/master/examples/nomad-consul-ami readme
+      "git clone --branch ${var.terraform_aws_nomad_version} https://github.com/hashicorp/terraform-aws-nomad.git /tmp/terraform-aws-nomad",
```

Comment:
I really hate that `--branch` also means "tag", too ☹️

Reply:
I think it would make more logical sense to split this up differently.

~~The `yum install` bit should be its own provisioner that runs first. You're essentially laying the foundation for what comes later.~~ Actually, this first provisioner should be the `setup_amazon-linux-2.sh` script, which otherwise appears unused. It even installs `git`, so this one here is redundant.

Then, this `git clone` bit should be moved into the `setup_nomad_consul.sh` script so all the related stuff is in one place. It's much easier to reason about things when you're reading that script if you know that there's not some extra bit floating around in the Packer template that you need to be aware of.

Reply:
yeah, turns out the setup-al2 script is only for
https://github.com/hashicorp/terraform-aws-nomad/blob/master/examples/nomad-consul-ami/nomad-consul-docker.json
not
https://github.com/hashicorp/terraform-aws-nomad/blob/master/examples/nomad-consul-ami/nomad-consul.json

so i'll just delete it and follow through with your other recommendation.

again, much of these weird choices are... well, not mine :)